docs/cookbook/creating-custom-schema-fields.md

```diff
@@ -0,0 +1,999 @@
+
+In this recipe, we are going to build two different custom schema fields. The first will be a simple set of `string` and `integer` fields to collect grade information. The second will be a gradient background creator with `integer` and `color` fields, within an `array`. Overall, the steps to building both will be almost identical. However, we are going to build the grade field as part of a piece and the gradient as a stand-alone module. This will demonstrate the changes you need to make when building your custom schema field in a module that doesn't extend another core module and changes that need to be made to `object` and `array` schema input fields when you use them this way.
+
+## Structure of a schema field
+
+A schema input field has two parts: a server-side part and a browser-side part.
+
+The server-side code handles registering the schema field type, plus sanitizing and storing the data in the database. Most of this code will be the same for each custom field. We can take advantage of the same code that the Apostrophe core uses for sanitization and storage.
+
+The browser-side code provides the admin UI and is where we will take advantage of the built-in fields to lighten the programming load. Apostrophe has two Vue components and two mixin files to assist with schema creation. The first component is the `AposInputWrapper`. This component helps with field validation and appearance. The other is the `AposSchema` component. This Vue component parses through a supplied schema of fields and implements the v-model that allows two-way data binding between a value in our template and a value in our data properties. Apostrophe provides the `AposInputMixin` mixin to facilitate passing props, display modifications, and event listeners. The `AposInputChoicesMixin` file assists in creating dynamic and static choices if you are using the `checkboxes`, `radio`, or `select` fields in your custom schema.
+
+## Building the custom grade field type
+![Bo Donkey's grades recorded in the new custom field](../.vuepress/public/images/recipes/custom-schema-grades.png)
+This custom field makes it easier for our CMS user, a teacher, to enter the grades for a student. Each set of grades is stored as a piece with the student's name as the `Title` field. Our custom schema makes it easy for the teacher to keep the mid-term and final grades organized for each class. Obviously, this example is overly simplistic. We could just as easily have added a `string` and two `integer` fields into an `array` field to achieve almost the same thing.
+
+### Implementing the server-side code
+For this example, we are creating a new `piece-type` module.
+
+<AposCodeBlock>
+
+```javascript
+module.exports = {
```

Comment:
These pieces aren't individual grade entries, since they contain an array of them. it's unclear what the piece type is for. The story might make more sense if the array of grades were attached to a `student` piece.

Reply:
I decided on the student piece.